internal/query/rolebindings.go

```diff
@@ -0,0 +1,624 @@
+		}
+
+		// process role relationships
+		roleID, err := gidx.Parse(rel.Subject.Object.ObjectId)
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.RoleBinding{}, err
+		}
+
+		dbRole, err := e.store.GetRoleByID(ctx, roleID)
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.RoleBinding{}, err
+		}
+
+		rb.Role = types.Role{
+			ID:   roleID,
+			Name: dbRole.Name,
+		}
+	}
+
+	return rb, nil
+}
+
+func (e *engine) CreateRoleBinding(ctx context.Context, resource, roleResource types.Resource, subjects []types.RoleBindingSubject) (types.RoleBinding, error) {
+	ctx, span := e.tracer.Start(
+		ctx, "engine.CreateRoleBinding",
+		trace.WithAttributes(
+			attribute.Stringer("role_id", roleResource.ID),
+			attribute.Stringer("resource_id", resource.ID),
+		),
+	)
+	defer span.End()
+
+	if err := e.isRoleBindable(ctx, roleResource, resource); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	dbrole, err := e.store.GetRoleByID(ctx, roleResource.ID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	role := types.Role{
+		ID:   roleResource.ID,
+		Name: dbrole.Name,
+	}
+
+	rbResourceType, ok := e.schemaTypeMap[e.rbac.RoleBindingResource.Name]
+	if !ok {
+		return types.RoleBinding{}, fmt.Errorf(
+			"%w: invalid role-binding resource type: %s",
+			ErrInvalidType, e.rbac.RoleBindingResource.Name,
+		)
+	}
+
+	rb := newRoleBindingWithPrefix(rbResourceType.IDPrefix, role)
+	roleRel := e.rolebindingRoleRelationship(role.ID.String(), rb.ID.String())
+
+	grantRel, err := e.rolebindingGrantResourceRelationship(resource, rb.ID.String())
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	updates := []*pb.RelationshipUpdate{
+		{
+			Operation:    pb.RelationshipUpdate_OPERATION_TOUCH,
+			Relationship: roleRel,
+		},
+		{
+			Operation:    pb.RelationshipUpdate_OPERATION_TOUCH,
+			Relationship: grantRel,
+		},
+	}
+
+	subjUpdates := make([]*pb.RelationshipUpdate, len(subjects))
+	rb.Subjects = make([]types.RoleBindingSubject, len(subjects))
+
+	for i, subj := range subjects {
+		rel, err := e.rolebindingSubjectRelationship(subj.SubjectResource, rb.ID.String())
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return types.RoleBinding{}, err
+		}
+
+		rb.Subjects[i] = subj
+		subjUpdates[i] = &pb.RelationshipUpdate{
+			Operation:    pb.RelationshipUpdate_OPERATION_TOUCH,
+			Relationship: rel,
+		}
+	}
+
+	if _, err := e.client.WriteRelationships(ctx, &pb.WriteRelationshipsRequest{
+		Updates: append(updates, subjUpdates...),
+	}); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
```

Comment:
Super important: You'll want to update the zedtoken for the resource that has the role binding so the binding takes effect immediately, otherwise there will be a substantial delay. Use `e.upsertZedToken(ctx, resource, zedToken)`.